interfaces/cython/cantera/test/test_composite.py

```diff
@@ -116,6 +117,95 @@ def check(a, b):
                     raise TypeError(msg) from inst
 
 
+class TestPickle(utilities.CanteraTest):
+
+    def test_pickle_gas(self):
+        gas = ct.Solution("h2o2.yaml", transport_model=None)
+        gas.TPX = 500, 500000, "H2:.75,O2:.25"
+        with open("gas.pkl", "wb") as pkl:
+            pickle.dump(gas, pkl)
+
+        with open("gas.pkl", "rb") as pkl:
+            gas2 = pickle.load(pkl)
+        self.assertEqual(gas.T, gas2.T)
+        self.assertEqual(gas.P, gas2.P)
+        self.assertTrue(np.allclose(gas.X, gas2.X))
+
+        self.assertEqual(gas2.transport_model, "None")
+
+    def test_pickle_gas_with_transport(self):
+        gas = ct.Solution("h2o2.yaml")
+        gas.TPX = 500, 500000, "H2:.75,O2:.25"
+        gas.transport_model = "Multi"
+        with open("gas.pkl", "wb") as pkl:
+            pickle.dump(gas, pkl)
+
+        with open("gas.pkl", "rb") as pkl:
+            gas2 = pickle.load(pkl)
+        self.assertEqual(gas.T, gas2.T)
+        self.assertEqual(gas.P, gas2.P)
+        self.assertTrue(np.allclose(gas.X, gas2.X))
+
+        self.assertEqual(gas2.transport_model, "Multi")
+
+    def test_pickle_interface(self):
+        gas = ct.Solution("diamond.yaml", "gas")
+        solid = ct.Solution("diamond.yaml", "diamond")
+        interface = ct.Interface("diamond.yaml", "diamond_100", (gas, solid))
+
+        with self.assertRaises(NotImplementedError):
+            with open("interface.pkl", "wb") as pkl:
+                pickle.dump(interface, pkl)
+
+
+class TestEmptyThermoPhase(utilities.CanteraTest):
+    """ Test empty Solution object """
+    @classmethod
+    def setUpClass(cls):
+        utilities.CanteraTest.setUpClass()
+        cls.gas = ct.ThermoPhase()
+
+    def test_empty_report(self):
+        with self.assertRaisesRegex(ct.CanteraError, "NotImplementedError"):
+            self.gas()
+
+    def test_empty_TPX(self):
+        with self.assertRaisesRegex(ct.CanteraError, "NotImplementedError"):
+            self.gas.TPX = 300, ct.one_atm, ""
+
+    def test_empty_equilibrate(self):
+        with self.assertRaisesRegex(ct.CanteraError, "NotImplementedError"):
+            self.gas.equilibrate("TP")
+
+
+class TestEmptySolution(TestEmptyThermoPhase):
+    """ Test empty Solution object """
+    @classmethod
+    def setUpClass(cls):
+        utilities.CanteraTest.setUpClass()
+        cls.gas = ct.Solution()
+
+    def test_empty_composite(self):
+        self.assertEqual(self.gas.thermo_model, "None")
+        self.assertEqual(self.gas.composite, ("None", "None", "None"))
+
+    def test_empty_kinetics(self):
+        kin = ct.Kinetics()
+        # an empty kinetics object is functional
```

Comment:
Shouldn't this use the `self.gas` object created in `setUpClass`, rather than create a new empty `Kinetics` object? That would seem more consistent with the name of the `TestEmptySolution` class. Ditto for `test_empty_transport` below.

Reply:
No. The `setUpClass` is used to override the `setUpClass` of the parent class so some tests can be 'recycled'. The other tests are added here as they only need to be run once. Tests are changed also, as empty `Kinetics` and `Transport` are no longer allowed.